firestore/bulkwriter.go

```diff
@@ -0,0 +1,521 @@
+	bwr *pb.BatchWriteRequest // request to send to the service
+	bwj []bulkWriterJob       // corresponding jobs to return response
+}
+
+type bulkWriterRequestBatch struct {
+	isFlushing bool              // flag that indicates that BulkWriter needs to schedule a flush of the queue
+	bwb        []bulkWriterBatch // all the bulkWriterBatch objects to schedule
+}
+
+type bulkWriterStats int
+
+const (
+	writesProvided bulkWriterStats = iota
+	writesSent
+	writesReceived
+)
+
+// The BulkWriterStatus provides details about the BulkWriter, including the
+// number of writes processed, the number of requests sent to the service, and
+// the number of writes in the queue.
+type BulkWriterStatus struct {
+	WritesProvidedCount int  // number of write requests provided by caller
+	IsOpen              bool // whether this BulkWriter is open or closed
+	WritesSentCount     int  // number of requests sent to the service
+	WritesReceivedCount int  // number of WriteResults received from the service
+}
+
+// A BulkWriter allows multiple document writes in parallel. The BulkWriter
+// submits document writes in maximum batches of 20 writes per request. Each
+// request can contain many different document writes: create, delete, update,
+// and set are all supported. Only one operation per document is allowed.
+// Each call to Create, Update, Set, and Delete can return a value and error.
+type BulkWriter struct {
+	database        string                      // the database as resource name: projects/[PROJECT]/databases/[DATABASE]
+	reqs            int64                       // total number of requests sent
+	start           time.Time                   // when this BulkWriter was started
+	vc              *vkit.Client                // internal client
+	isOpen          bool                        // signal that the BulkWriter is closed
+	startFlush      chan struct{}               // signal the beginning of flushing operation
+	isFlushing      bool                        // determines whether we're in a flush state
+	endFlush        chan struct{}               // signal the end of flushing
+	received        chan bulkWriterJob          // incoming write requests
+	scheduled       chan bulkWriterRequestBatch // scheduled requests
+	sendingQueue    []bulkWriterJob             // queue of requests to send
+	backlogQueue    []bulkWriterJob             // queue of requests to store in memory during flush operation
+	maxOpsPerSecond int                         // number of requests that can be sent per second.
+	docUpdatePaths  sync.Map                    // document paths with corresponding writes in the queue.
+	waitTime        float64                     // time to wait in between requests; increase exponentially
+	writesLog       sync.Map                    // stores writes provided, sent, and received
+	cancel          context.CancelFunc          // context to send cancel message
+	mu              sync.Mutex                  // locks concurrent access
```

Comment:
nit: Anytime a mu is used in a struct seperate with whitespace the mutext and under it put the items the mutex is protecting. This makes it clear which fields are gauraded by what mutex.